[WIP] What: Cloud Native Guide Posts Issue #961 (#997)
cncf/toc: process/project_milestones.md

## Sandbox

Some milestones may not apply to all projects and should be leveraged as guiding points. Milestones are described as the desired state with examples for how to achieve this.
* Project has the basics of security underway
  * Completion of the [self-assessment](https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md)
* Project holds community meetings or active discussions
  * Zoom with meeting notes occurring more than twice a quarter (or community.cncf.io!) 
  * Use of GitHub discussions, issues, or a group messaging service (slack, wechat, etc.)
* Project guides community contributions
  * PRs from community members receive timely feedback and discussion from the maintainers
* Project works with an angel adopter
  * Partners closely with the project to address organizational needs of the project early on
* Project begins solidifying a versioning schema and release cadence
  * A roadmap clearly defines initiatives that align with releases

>   * Releases begin to have establish a pattern of regularity

Reviewer: ```suggestion
  * Releases begin to establish a pattern of regularity
```

> * Project has engaged a Technical Advisory Group for feedback on some portion of the project relevant to the domain

Reviewer: ```suggestion
* Project has engaged a [Technical Advisory Group](/tags/README.md) for feedback on some portion of the project relevant to the domain
```

> * Project has a clearly discoverable governance doc that covers the basics of decision making to include the commit bit

Reviewer: ```suggestion
* Project has a clearly discoverable governance doc that covers the basics of decision making and how to earn the commit bit
```

Reviewer 2: "the commit bit" is probably not a well understood term.  I don't know if the CNCF has a defined term (["maintainer" is definitely overloaded](https://github.com/cncf/foundation/issues/433)), but I think "how to earn permission to approve pull requests" or something similar would help.

(I assume the SCM for CNCF projects does not _have_ to include GitHub and a more inclusive terms than Pull Requests might be appropriate)

Reviewer 3: I've been thinking about Roadmaps. In the past I've been in favour of them as a requirement but I think I'm changing my mind because I'm not convinced they are as valuable as we'd like to think. 

Background / food for thought: the Linux kernel does not have a published roadmap. Everyone involved knows what's going through mailing list discussions.

When we say things like a roadmap that's aligned with releases, combined with a regular cadence of releases, this feels dangerously close to requiring open source projects to make commitments that end users might come to think they can rely on.

The main concern I have is that it doesn't seem very aligned with the ethos of open source to require developers to sign up to fixed commitments. Sure, in our community most people are being paid to work on projects but not everyone is. And if someone from Company A makes a commitment to deliver a particular feature in a certain release, but then Company A decides they should spend their time elsewhere, who from the project is supposed to pick up the slack to make sure the commitment is reached? 

It might work for very large projects, but the more I think about it, the more it seems like a recipe for burnout amongst contributors, and disappointment amongst end users. 

Projects generally use issues to communicate what features and fixes people are working on. A feature that meets some set of criteria (up to the project) makes it into the next release. Isn't that sufficient? 

And if we require more, isn't that rather overstepping the idea that we don't impose any particular governance on projects?

Reviewer 4: +100 - we really need to stop turning CNCF into a giant bureaucracy that treats all projects the same.  This kills innovation stone dead and will put so many people off.  Also, please stop imposing requirements on project maintainers that add a lot of work and will lead to burnout, without a very strong compensating reason and balance

Author: @lizrice Thanks for bringing this up, I dont disagree. We currently have roadmaps integrated through various portions of the lifecycle so we'll need to evaluate what the value is of these in both understanding the project's longer goals as well as the direction they are heading, however they are always point in time for evaluations along the lifecycle. 

Your point on perceived commitment is spot on as well, and I'm not confident "messaging" here will resolve this disconnect. We'll need to consider more about how we can view the direction of the project at an evaluation, without dictating how the project expresses that.

Author: @monadic Thank you for the feedback.  This PR on milestones does not establish new or additional requirements on projects, it is designed to serve as ` guiding points along their [project's] journey — from sandbox to beyond graduation` which they `may leverage`.  However, as @craigbox mentioned later, these may be seen as requirements. This is something will need to rework in this PR to avoid.
Regarding burnout, agreed. The CNCF lifecycle is structured to evaluate projects for stability, maturity, and adoption potential (to varying degrees) over the years, as we've learned more about project success, we're beginning to distill steps projects have taken along their path to success. Not as requirements, more as a catalog of what has and hasnt worked and more context around why.  

Would providing more reasoning/clarity/what-it-does assist with the milestones? Such that as a project maintainer looking to move to incubation from sandbox (for instance) has ample information about how these may be beneficial   to decide if documenting a roadmap is worthwhile for their project, or perhaps including a vision/direction, or neither, works for their needs?

Reviewer 5: I've suggested different language for this point.

> * TBD - add more!

Reviewer: ```suggestion
* Project has an Adopters file and starts capturing public production users and use cases
* TBD - add more!
```

Author: Including this but adjusting language to reflect more project driven implementation with examples of "how"

## Incubation

Incubation milestones are guiding points to highlight success areas a maturing project can accomplish before Graduation. Not all Incubating projects will achieve every milestone, however many successful graduated projects have demonstrated these milestones prior to their Graduation. How each project achieves some or all of these milestones will vary widely. Projects do not need to meet this in order to graduate, rather these assist projects in meeting the Graduation criteria.

> * Project has a robust set of governance documentation that defines how the direction of the project is decided upon and managed which considers company diversity in decision making

Reviewer: ```suggestion
* Project has a robust set of governance documentation that defines how the direction of the project is decided upon and managed which considers company diversity in decision making. TAG Contributor Strategy has many great [governance templates](https://github.com/cncf/project-template)
```

* Project can demonstrate its application, practice, and adjustments to its governance documentation as a result of regular project operations
* Project has a robust and mature security posture as evidenced by
  * Threat model
  * [Joint review](https://github.com/cncf/tag-security/tree/main/assessments#components-of-the-security-review-package)

>   * First passing then Silver OpenSSF Best Practice Badge 

Reviewer: This is currently different than graduation requirements. I would keep it consistent across both places https://github.com/cncf/toc/blob/main/process/graduation_criteria.md#graduation-stage

Author: @xmulligan it is different from graduation requirements. Perhaps this is better suited as a split? Achieve Passing during incubation (to align with graduation) and then Silver post graduation?  these are not requirements and are optional for project (except where they exist as requirements). Silver has (IMO) reasonable and (in all likelihood)  a fair amount of requirements beyond passing that are already met by highly mature projects.

What do you think of this change? breaking it out and moving Silver to after graduation?

Reviewer 6: As a maintainer, I think this makes sense to still have "goals"/"points for improvement" after graduating as that is not the end of the governance story for a project

* Project documentation includes getting started guides, operating/administration instruction, security call-outs, and other core elements necessary to ease adoption.

> * Project provides scalability benchmarks and sets expectations around scaling

Reviewer: This is unclear to me. Do we have examples or resources we could provide here?

Author: This would be a new discipline (and should include performance with scalability) - there are a few projects that have done comparisons of performance between themselves and other projects and products.  This would drop the comparison portion and rather focus on the recommended architecture, a few deployment options (single cluster, multi-cluster, multi-cloud), how it scales (increasing nodes, increased traffic, load, or processes, expected completion time, etc.) It may also be as simple as expressing the CPU utilization of running the project depending on various configurations.  The goal would be to set expectations with potential users/adopters that may also be beneficial in understand if the project is performing optimally in their environment or if there are other factors contributing to performance or scaling degradation.

Reviewer 6: As a maintainer, I believe it's good to also define when this is _not_ a requirement. For example, KEDA handles autoscaling but there is no aspect of scalability/performance benchmark as we're just building on top of KEDA with an additional operator.

You can always make the case that this is still required; but for some projects I believe it is less beneficial or the impact on a project would be too big.

Again for KEDA, we'd need to spin up big (expensive) clusters with sample workloads (1000+) to do benchmarking of something that is just built on top of Kubernetes' building blocks (HPA / metric servers)

* Project captures use cases to showcase how it can be best used to solve common problems.
  * These may leverage observations and concerns from adopters
* Project contribution and community documentation clearly defines expectations when contributing (to include non-code contributions).

> * Project defines a process to “build the bench” of maintainers and project leaders such as a Contribution Ladder which considers succession planning

Reviewer: ```suggestion
* Project defines a process to “build the bench” of maintainers and project leaders such as a Contribution Ladder which considers succession planning and getting maintainers from multiple organizations
```

# Cloud Native Milestones

> Milestones are intended to outline checkpoints for projects, that have assisted past projects as they grew and mature in the cloud native ecosystem. They are NOT requirements for moving between levels, please refer to [the graduation criteria](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md) for requirements. 

Reviewer 7: ```suggestion
Milestones are intended to outline checkpoints for projects, that have assisted past projects as they grew and matured in the cloud native ecosystem. They are NOT requirements for moving between levels, please refer to [the graduation criteria](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md) for requirements. 
```

Suggesting past tense for "mature" considering the sentence's past tense of "grew".

* Project has the basics of security underway
  * Completion of the [self-assessment](https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md)
* Project holds community meetings or active discussions
  * Zoom with meeting notes occurring more than twice a quarter (or community.cncf.io!) 
  * Use of GitHub discussions, issues, or a group messaging service (slack, wechat, etc.)
* Project guides community contributions
  * PRs from community members receive timely feedback and discussion from the maintainers
* Project works with an angel adopter
  * Partners closely with the project to address organizational needs of the project early on
* Project has a second organization contributing
  * May begin path to maintainership
* Project begins solidifying a versioning schema and release cadence
  * A roadmap or other document clearly defines project direct or initiatives that may align with releases
  * Releases begin to establish a pattern of regularity
* Project has engaged a [Technical Advisory Group](/tags/README.md) for feedback on some portion of the project relevant to the domain

> * Project has a clearly discoverable governance doc that covers the basics of decision making and how to earn permissions to approve pull request, sometimes referred to as "the commit bit"

Reviewer 5: ```suggestion
* Project has a clearly discoverable governance doc that covers the basics of decision making, including who approves PRs and how committers/maintainers are chosen
```

* Project has an angel adopter that assists in stability and production runtime or deployment
* Project has 3 angel adopters

>   * May be from different industry verticals

Reviewer 5: Not quite understanding the language around "angel adopters", can this explain better?

  * May have differing use cases
* TBD - add more!

## Incubation

Incubation milestones are guiding points to highlight success areas a maturing project can accomplish before Graduation. They are not requirements. Not all Incubating projects will achieve every milestone, however many successful graduated projects have demonstrated these milestones prior to their Graduation. How each project achieves some or all of these milestones will vary widely. Projects do not need to meet this in order to graduate, rather these assist projects in meeting the Graduation criteria.

> * Project has a robust set of governance documentation that defines how the direction of the project is decided upon and managed which considers company diversity in decision making. [TAG Contributor Strategy](https://github.com/cncf/tag-contributor-strategy) has many great [governance templates](https://github.com/cncf/project-template).

Reviewer 5: ```suggestion
* Project is evolving and developing a robust set of governance documentation that defines how the project is managed, and its roadmap is determined, which considers maintainer turnover and company diversity. [TAG Contributor Strategy](https://github.com/cncf/tag-contributor-strategy) has many great [governance templates](https://github.com/cncf/project-template).
```

* Project ensures all sub-projects are clearly defined in their maturity level, whether they ship with the core project or can be used independently but is subject to the same governance processes
* Project refreshes their roadmap planning with a reasonable cadence
* Project has engaged with a few Technical Advisory Groups to ensure they are considering best practices across multiple technical and governance facets
* Project has defined a process for managing the conduct of its community, either establishing its own group to do this, leveraging the existing leadership, or deferring to the CNCF
* Project maintains a dependency graph so adopters fully understand the complexity and risk of use
* TBD - add more!

## Graduation

Graduation milestones are guiding points to highlight continuing success areas for highly mature project after they graduate. They are not requirements. Not all Graduated projects will achieve or perform these milestones, however many robust and mature projects will adopt these milestones as "regular business" to maintain the health and growth of the project and its community.

> * Project reviews it's governance process annually for improvements to improve clarity, consistency, and inclusivity

Reviewer 5: ```suggestion
* Project regularly updates its governance documentation to accurately reflect changes to processes, staffing, roles, and activities
```

We're already giving Incubating projects the full milestone of "complete governance", which means that all we can ask graduated projects to do is to keep that up to date.  No?

> * Project has a well established and exercised contributor growth ladder or other community construct to "build the bench" of maintainers or project leaders

Reviewer 5: We already ask for this at incubating.  Would it be worth asking a project to have demonstrated turnover?

Author: Done: https://github.com/cncf/toc/pull/1020

## Sandbox

Some milestones may not apply to all projects and should be leveraged as guiding points. Milestones are described as the desired state with examples for how to achieve this. They are not requirements.
* Project has the basics of security underway
  * Completion of the [self-assessment](https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md)
* Project holds community meetings or active discussions
  * Zoom with meeting notes occurring more than twice a quarter (or community.cncf.io!) 
  * Use of GitHub discussions, issues, or a group messaging service (slack, wechat, etc.)
* Project guides community contributions
  * PRs from community members receive timely feedback and discussion from the maintainers

> * Project works with an angel adopter

Reviewer 2: Duplicated at line 26?

* Project is evolving and developing a robust set of governance documentation that defines how the project is managed, and its roadmap is determined, which considers maintainer turnover and company diversity. [TAG Contributor Strategy](https://github.com/cncf/tag-contributor-strategy) has many great [governance templates](https://github.com/cncf/project-template).
* Project can demonstrate its application, practice, and adjustments to its governance documentation as a result of regular project operations
* Project has a robust and mature security posture which could be achieved by one or more of the following:
  * Threat model
  * [Joint review](https://github.com/cncf/tag-security/tree/main/assessments#components-of-the-security-review-package)
  * [Passing level](https://bestpractices.coreinfrastructure.org/en/criteria/0) of the [OpenSSF Best Practice Badge](https://bestpractices.coreinfrastructure.org/)

>     * Note: Completion and maintenance of **a** OpenSSF Best Practice Badge is [a requirement for graduation](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md#graduation-stage)

Reviewer 2: `**an**` 😉

* Project has engaged with a few Technical Advisory Groups to ensure they are considering best practices across multiple technical and governance facets
* Project has defined a process for managing the conduct of its community, either establishing its own group to do this, leveraging the existing leadership, or deferring to the CNCF
* Project maintains a dependency graph so adopters fully understand the complexity and risk of use
* Project has 3 or more adopters using the project in production to support completion of Due Diligence and adopter interviews

> * Project plans activities to increase project adoption trajectory 3 adopters

Reviewer 3: ```suggestion
* Project plans activities to increase project adoption 
```

> * Project has more than 5 regular contributors, more than half of which are from different organizations

Reviewer 3: This one worries me A LOT. It clearly incentivizes projects and companies to do the wrong thing, in order to meet these "milestones".  

For example, fearing that they are being seen to have "too many" contributions going into the open source project, the major backer of a project decides to hold their engineering efforts back into closed-source add-ons. Or use outsourcing companies and partners rather than hire people to work full-time on the project. Or decide not to contribute a project to the CNCF because it's going to be such an uphill struggle for the project to see itself as successful according to these criteria. 

None of these outcomes are good for anyone in our community. 

(Also, why arbitrarily 5? And what does "regular" mean? We could spend time debating these points - but that is time that could be better spent on actually progressing / supporting projects.)

Author: These are good counterpoints @lizrice . We also don't want to have a single entity (company or organization or individual) be the majority decision maker of a projects technical direction, if that entity is acquired or disappears, the project can be left out to dry, and in some case unable to recover because the primary pool of contribution has evaporated.

@raravena80 @lizrice - How do we strike a balance in these milestones — as guiding points projects _could_ leverage to in order to become more resilient, mature, and widely adopted versions — that could help projects avoid the scenarios you present as well as a increase the likelihood of project's survivability with primary entity's complete absence?  What _actionable_ goals/ideals/structure/governance could be provided for projects to increase their resilience and survival likelihood in the event a core contributing entity ceases to be or continue? 

Ricardo mentioned open and flexible language here, would the following adjustments meet the intent?

```suggestion
* Project has multiple contributors who regularly participate in the growth and development of the project in accordance with the projects synchronous and asynchronous cadence
  * Project contribution governance defines timeliness and methods for recovering roles and permissions of contributors that drop from regular participation as _defined by the project_ (roles such as issue triage, community management, PR reviewer, maintainer, release lead, etc. and regularity such as participation in two of three releases, facilitation of 4 community meetups annually, etc.)
```

Reviewer 5: I can't say I'm keen on removing specifics from this.  Folks in the community manager role for a project like having clear targets.  Once a project gets to applying Graduated, if they don't have at least five regular contributors, the application isn't likely to go well.  

Note that these are milestones, not updates to the specific requirements.

Reviewer 8: 👍  I believe all these can be covered in the [March 21st TOC meeting](https://docs.google.com/document/d/1jpoKT12jf2jTf-2EJSAl4iTdA7Aoj_uiI19qIaECNFc/edit#heading=h.8q8nrqluehva).
